src/leafletwidget/widget.ts

```diff
@@ -80,28 +77,51 @@ class MapWidget extends Widget {
    */
   protected onUpdateRequest(msg: Message): void {
     this.title.text = this._context.path.split('/').pop();
-    let cm = this._context.contentsModel;
-    if (cm === null) {
+    if (!this.isAttached) {
       return;
     }
-    let content = JSON.parse(this._context.model.toString());
+    let contentString = this._context.model.toString();
+    if (contentString === this._geojsonString) {
+      return;
+    }
+
+    // we're attached to the DOM and have new layer content now
     if (this._geojsonLayer) {
       this._map.removeLayer(this._geojsonLayer);
     }
-    this._geojsonLayer = leaflet.geoJson(content, {
-      pointToLayer: function (feature, latlng) {
-          return leaflet.circleMarker(latlng);
-      }
-    });
-    this._map.addLayer(this._geojsonLayer);
-    this._map.fitBounds(this._geojsonLayer.getBounds());
+    this._geojsonString = contentString;
+    this._geojsonLayer = null;
+    if (contentString) {
+      let content = JSON.parse(contentString);
+      this._geojsonLayer = leaflet.geoJson(content, {
+        pointToLayer: function (feature, latlng) {
+            return leaflet.circleMarker(latlng);
+        }
+      });
+      this._map.addLayer(this._geojsonLayer);
+      this._map.fitBounds(this._geojsonLayer.getBounds());
+    }
   }
 
   /**
    * A message handler invoked on a 'resize' message.
    */
-  onResize() {
+  onResize(msg: ResizeMessage) {
+    // Since we know the size from the resize message, we manually
+    // define getSize() so that it does not have to do a DOM read.
+    this._map.getSize = () => {
+      if (msg.width === -1 || msg.height === -1) {
+        return (this._map as any).prototype.getSize();
+      } else {
+        let size: any = new leaflet.Point(msg.width, msg.height);
+        (this._map as any)._size = size;
+        return size.clone();
+      }
+    };
```

Comment:
Ah. Cool. Since the author recommends patching `getSize` maybe something like this would work (pseudo code):

```
// `this` is a Widget
this._map.getSize = () => {
  if (this._mapWidth !== -1 && this._mapHeight !== -1) {
    return L.point(this._mapWidth, this._mapHeight);
  }
  return L.point(this._mapNode.offsetWidth, this._mapNode.offsetHeight);
};
```

And update `_mapWidth` and `_mapHeight` from the resize message handler.